File: 외판원순환문제2.py
from itertools import permutations
import sys
input = sys.stdin.readline
n = int(input())
w = [list(map(int,input().split())) for _ in range(n)]
perm = list(range(n))

INT_MAX =2147483647
ans = INT_MAX

for p in list(permutations(perm)):
  if w[p[-1]][p[0]] == 0 : continue
  
  isPossible = True
  costs = 0
  for i in range(n-1):
    if w[p[i]][p[i+1]] == 0 :
      isPossible = False
      break
    else : 
      costs+= w[p[i]][p[i+1]]
      if costs >= ans :
        isPossible=False
        break
  if isPossible == False : continue
  costs+=w[p[-1]][p[0]]
  ans = min(ans,costs)
print(ans)

File: 외판원순환문제2.cpp
#include <iostream>
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

const int INT_MAX =2147483647;
int w[11][11];

int main(){
  int n ; 
  cin >>n;
  vector<int> perm(n);
  
  for(int i=0; i<n; i++){
    for(int j=0; j<n; j++){
      cin >> w[i][j];
    }
  }
  for(int i=0; i<n; i++){
    perm[i] = i;
  }
  int ans = INT_MAX;
  do{
    if(perm[0] != 0) break;
    bool isPossible = true;
    int costs = 0; 
    for(int i=0; i<n-1; i++){
      if(w[perm[i]][perm[i+1]] == 0) isPossible = false;
      else costs+=w[perm[i]][perm[i+1]];
    }
    if(isPossible&& w[perm[n-1]][perm[0]] != 0){
      costs += w[perm[n-1]][perm[0]];
      ans = min(ans,costs);
    }
  }while(next_permutation(perm.begin()+1,perm.end()));
  cout<<ans<<'\n';
}